Imports

In [1]:
import pandas as pd
import numpy as np
pd.options.display.max_rows = 320
df_scrape = pd.read_csv("./22_10_22_hls_people_hist_hub_V2.csv", sep=";")
df_scrape.drop("Unnamed: 10",axis=1, inplace=True)

# replace empty strings and zero width spaces with NaN
df_scrape.replace(r'^\s*$', np.nan, regex=True, inplace=True)
df_scrape.replace(r'^\u200b*$', np.nan, regex=True, inplace=True)

df_scrape.head(10)

HLS_ID                                            HLS_URL       first_name  \
0   25265  http://hls-dhs-dss.ch/de/articles/025265/2001-...    Johann Ludwig   
1   27932  http://hls-dhs-dss.ch/de/articles/027932/2002-...   Carlo Emanuele   
2    4928  http://hls-dhs-dss.ch/de/articles/004928/2002-...   Albert von der   
3   11462  http://hls-dhs-dss.ch/de/articles/011462/2001-...         Johannes   
4    8115  http://hls-dhs-dss.ch/de/articles/008115/2001-...            Alvar   
5   18417  http://hls-dhs-dss.ch/de/articles/018417/2001-...  Johann III. von   
6   25265  http://hls-dhs-dss.ch/de/articles/025265/2001-...    Johann Ludwig   
7   12040  http://hls-dhs-dss.ch/de/articles/012040/2000-...         Nesa von   
8   12072  http://hls-dhs-dss.ch/de/articles/012072/2017-...      Lüthold von   
9   32200  http://hls-dhs-dss.ch/de/articles/032200/2001-...    Rudolf I. von   

  last_name birth_date death_date  \
0   A Musis   8.8.1599  10.6.1662   
1   A Prato   7.4.1895  17.9.1968   
2        Aa  19.1.1894  10.7.1978   
3       Aal    um 1500  28.5.1551   
4     Aalto   3.2.1898  11.5.1976   
5   Aarberg    um 1410       1497   
6   A Musis   8.8.1599  10.6.1662   
7   Aarberg       1398   vor 1438   
8   Aarburg       1190   7.6.1213   
9   Aarburg       1296  16.7.1339   

                                                text   published  \
0  8.8.1599 Pfullendorf (Baden, D), 10.6.1662 Sol...  10.05.2001   
1  7.4.1895 Trient, 17.9.1968 Genf, ursprünglich ...  21.08.2002   
2  19.1.1894 Vevey, 10.7.1978 Lausanne, reformier...  11.06.2002   
3  um 1500 Bremgarten (AG), 28.5.1551 Solothurn, ...  24.01.2001   
4  3.2.1898 Kuortane (Finnland), 11.5.1976 Helsin...  15.05.2001   
5  um 1410, 1497. Sohn des Wilhelm, Herrn von Val...  15.05.2001   
6  8.8.1599 Pfullendorf (Baden, D), 10.6.1662 Sol...  10.05.2001   
7  Erstmals erwähnt 1398, vor 1438 Basel. Aus der...  18.10.2000   
8  1190 erstmals bezeugt, 7.6.1213. Aus dem Gesch...  08.05.2017   
9  1296 als Student in Bologna erstmals erwähnt, ...  24.01.2001   

                   author                translator   hhb_ids    hhb_forename  \
0              Urban Fink                       NaN  11502881   Johann Ludwig   
1           Mauro Cerutti  Christoph Neuenschwander  11519981  Carlo Emanuele   
2          Gilbert Marion               Peter Opitz  11383091          Albert   
3          Rolf Max Kully                       NaN  11420099        Johannes   
4        Simona Martinoli              Sabine Kraut  11403170           Alvar   
5        Germain Hausmann           Gertraud Gamper  11466260          Johann   
6              Urban Fink                       NaN  11502881   Johann Ludwig   
7    Veronika Feller-Vest                       NaN  11423564            Nesa   
8            Romain Jurot         Ansgar Wildermann  11423816         Lüthold   
9  Franziska Hälg-Steffen                       NaN  11551625          Rudolf   

   hhb_surname hhb_sex hhb_birth_date hhb_death_date  hhb_relation_id  \
0      A Musis    male     1599-08-08     1662-06-10              NaN   
1      A Prato    male     1895-04-07     1968-09-17              NaN   
2   von der Aa    male     1894-01-19     1978-07-10              NaN   
3          Aal    male     1500-01-01     1551-05-28              NaN   
4        Aalto    male     1898-02-03     1976-05-11              NaN   
5  von Aarberg    male     1410-01-01     1497-01-01              NaN   
6      A Musis    male     1599-08-08     1662-06-10              NaN   
7  von Aarberg  female     1398-01-01     1438-01-01              NaN   
8  von Aarburg    male     1190-01-01     1213-06-07              NaN   
9  von Aarburg    male     1296-01-01     1339-07-16              NaN   

  hhb_relation_name hhb_relation_type  
0               NaN               NaN  
1               NaN               NaN  
2               NaN               NaN  
3               NaN               NaN  
4               NaN               NaN  
5               Na

## Clean Malformed HTML
https://hls-dhs-dss.ch/de/articles/019242/2007-08-09/


In [2]:
# get all records with missing first_name and last_name
df_scrape[df_scrape.first_name.isnull() & df_scrape.last_name.isnull()]

HLS_ID                                            HLS_URL first_name  \
11421   19242  http://hls-dhs-dss.ch/de/articles/019242/2007-...        NaN   

      last_name birth_date death_date  \
11421       NaN        NaN       1607   

                                                    text published  \
11421  1607 Therwil, katholisch, von Konstanz. Ca. 15...       NaN   

                 author translator   hhb_ids hhb_forename hhb_surname hhb_sex  \
11421  André Salvisberg        NaN  11472911      Abraham      Keller    male   

      hhb_birth_date hhb_death_date  hhb_relation_id hhb_relation_name  \
11421     1557-01-01     1607-01-01              NaN               NaN   

      hhb_relation_type  
11421               NaN

In [3]:
df_scrape.loc[11421, "first_name"] = "Abraham"
df_scrape.loc[11421, "last_name"] = "Keller"
df_scrape.loc[11421, "published"] = "09.08.2007"

## Checking for Null Values

In [4]:
df_scrape.notna().sum()

HLS_ID               25271
HLS_URL              25271
first_name           24803
last_name            25271
birth_date           24601
death_date           24641
text                 25271
published            25271
author               25271
translator            6588
hhb_ids              25205
hhb_forename         25198
hhb_surname          24742
hhb_sex              25198
hhb_birth_date       25132
hhb_death_date       24449
hhb_relation_id        102
hhb_relation_name      102
hhb_relation_type      102
dtype: int64

- first name, is clear because lots of records before 1000 didn't specify a first name
- translator, only specified if the text was translated to german (hence, fine)
- birth_date & death_date needs some investigation

In [5]:
len(df_scrape[df_scrape.birth_date.isnull() & df_scrape.hhb_birth_date.isnull()])

72

In [6]:
len(df_scrape[df_scrape.death_date.isnull() & df_scrape.hhb_death_date.isnull()])

466

In [7]:
len(df_scrape[df_scrape.birth_date.isnull() & df_scrape.hhb_birth_date.isnull() & df_scrape.death_date.isnull() & df_scrape.hhb_death_date.isnull()])

12

There are 72 birth dates and 466 death dates which we can not source from either hls or hist hub (hhb).
And there are 12 records where we can't source any date information from either source.

In [8]:
df_scrape[df_scrape.birth_date.isnull() & df_scrape.hhb_birth_date.isnull() & df_scrape.death_date.isnull() & df_scrape.hhb_death_date.isnull()]

HLS_ID                                            HLS_URL  first_name  \
1848    60641  http://hls-dhs-dss.ch/de/articles/060641/2022-...        Leny   
2031    59102  http://hls-dhs-dss.ch/de/articles/059102/2020-...   Henriette   
6450    10200  http://hls-dhs-dss.ch/de/articles/010200/2008-...         NaN   
7774    58885  http://hls-dhs-dss.ch/de/articles/058885/2020-...     Jrenäus   
10433   61175  http://hls-dhs-dss.ch/de/articles/061175/2022-...        Rosa   
12808   60087  http://hls-dhs-dss.ch/de/articles/060087/2021-...        Elsy   
14474   58877  http://hls-dhs-dss.ch/de/articles/058877/2020-...  Jean-Louis   
15367   58886  http://hls-dhs-dss.ch/de/articles/058886/2020-...     Theodor   
15440   58879  http://hls-dhs-dss.ch/de/articles/058879/2020-...    François   
16731   58875  http://hls-dhs-dss.ch/de/articles/058875/2020-...    Adèle de   
22334   58878  http://hls-dhs-dss.ch/de/articles/058878/2020-...       Marie   
22836   10332  http://hls-dhs-dss.ch/de/articles/010332/2014-...         NaN   

                  last_name birth_date death_date  \
1848                  Bider        NaN        NaN   
2031        Bitzius-Zeender        NaN        NaN   
6450       Felix und Regula        NaN        NaN   
7774               Gersdorf        NaN        NaN   
10433      Häcki-Feierabend        NaN        NaN   
12808             Leuzinger        NaN        NaN   
14474               Micheli        NaN        NaN   
15367                Müller        NaN        NaN   
15440                  Naef        NaN        NaN   
16731                Pierre        NaN        NaN   
22334     Tourte-Cherbuliez        NaN        NaN   
22836  Vadomar und Gundomad        NaN        NaN   

                                                    text   published  \
1848   ​8.11.1894 Langenbruck,​7.7.1919 Zürich, refor...  17.08.2022   
2031   ​8.8.1805 Bern, ​14.6.1872 Bern, reformiert, v...  08.04.2020   
6450   Das früheste Zeugnis der beiden Zürcher Stadth...  15.12.2008   
7774   ​1809 Monstab bei Altenburg (Thüringen), ​1860...  08.04.2020   
10433  ​5.1.1922 Engelberg, ​14.10.1996 Engelberg, ka...  26.07.2022   
12808  ​7.2.1910 Kilchberg (ZH), ​27.4.2010 Zürich, r...  24.08.2021   
14474  ​16.12.1812 Genf, ​30.3.1875 Genf, reformiert,...  08.04.2020   
15367  ​12.11.1790 Strelitz-Alt, ​23.7.1857 Hofwil, p...  08.04.2020   
15440  ​31.3.1825 Genf, ​4.11.1897 Genf, reformiert, ...  08.04.2020   
16731  ​1.4.1800 Neuenburg, ​5.8.1890 Neuenburg, refo...  08.04.2020   
22334  ​12.4.1793 Genf, ​19.8.1863 Plainpalais, refor...  08.04.2020   
22836  Vadomar (​nach 371) und Gundomad (357) waren a...  04.03.2014   

                    author           translator   hhb_ids      hhb_forename  \
1848          Tamara Suter                  NaN  18001995               NaN   
2031       Marianne Derron                  NaN       NaN               NaN   
6450          Hans Stadler                  NaN  11412818  Felix und Regula   
7774         Jesko Reiling                  NaN       NaN               NaN   
10433  Heidi Wernli Gasser                  NaN       NaN               NaN   
12808  Susanne Peter-Kubli                  NaN       NaN               NaN   
14474      Marianne Derron                  NaN       NaN               NaN   
15367       Michael Ruloff                  NaN       NaN               NaN   
15440      Marianne Derron                  NaN       NaN               NaN   
16731      Marianne Derron                  NaN       NaN               NaN   
22334      Marianne Derron                  NaN       NaN               NaN   
22836    Gabrielle Duchoud  Susanne Ritter-Lutz       NaN               NaN   

      hhb_surname hhb_sex hhb_birth_date hhb_death_date  hhb_relation_id  \
1848          NaN     NaN            NaN            NaN              NaN   
2031          NaN     NaN            NaN            NaN              NaN   
6450          NaN    male            NaN            NaN              NaN   
7774   

In [9]:
df_scrape[df_scrape.death_date.isnull()].iloc[0].HLS_URL

'http://hls-dhs-dss.ch/de/articles/012450/2001-06-05/'

## Trying to get value out of hls date fields.

In [10]:
# str extract with date format
df_scrape.birth_date.str.extract(r"((\d{1,2}\.\d{1,2}\.\d{3,4})|(\d{4}))")
# python str.extract regex or str.extractall regex


0           1     2
0        8.8.1599    8.8.1599   NaN
1        7.4.1895    7.4.1895   NaN
2       19.1.1894   19.1.1894   NaN
3            1500         NaN  1500
4        3.2.1898    3.2.1898   NaN
...           ...         ...   ...
25266   24.1.1887   24.1.1887   NaN
25267  11.11.1921  11.11.1921   NaN
25268    7.3.1304    7.3.1304   NaN
25269        1445         NaN  1445
25270   15.6.1558   15.6.1558   NaN

[25271 rows x 3 columns]

In [11]:
# Total Rows: (25271)
df_scrape.birth_date.isna().sum() + df_scrape.birth_date.notna().sum()

25271

In [12]:
df_scrape.birth_date.count()

24601

In [13]:
# Number of rows in birth_date being matched by regex: (24547)
df_scrape.birth_date.str.extract(r"((\d{1,2}\.\d{1,2}\.\d{3,4})|(\d{3,4}))")[0].notna().sum()

24547

The proof below shows that we indeed match the first year if specification year/year is hit:

In [14]:
df_scrape.birth_date[62:63]

62    610/611
Name: birth_date, dtype: object

In [15]:
df_scrape.birth_date.str.extract(r"((\d{1,2}\.\d{1,2}\.\d{3,4})|(\d{3,4}))")[62:63]

0    1    2
62  610  NaN  610

Let's continue

In [16]:
# Number of rows in birth_date NOT being matched by regex: (54)
df_scrape.birth_date.notna().sum() - df_scrape.birth_date.str.extract(r"((\d{1,2}\.\d{1,2}\.\d{3,4})|(\d{3,4}))")[0].notna().sum()

54

Let's investigate those 54 records, 48 of these records have a hhb_birth_date which covers those records.

In [17]:
df_scrape[df_scrape.birth_date.str.extract(r"((\d{1,2}\.\d{1,2}\.\d{3,4})|(\d{3,4}))")[0].isna() & df_scrape.birth_date.notna() & df_scrape.hhb_birth_date.isna()]

HLS_ID                                            HLS_URL  \
7769    12640  http://hls-dhs-dss.ch/de/articles/012640/2005-...   
11695   15548  http://hls-dhs-dss.ch/de/articles/015548/2006-...   
13809   58876  http://hls-dhs-dss.ch/de/articles/058876/2020-...   
14570   12835  http://hls-dhs-dss.ch/de/articles/012835/2008-...   
16168   48410  http://hls-dhs-dss.ch/de/articles/048410/2009-...   
18142   12921  http://hls-dhs-dss.ch/de/articles/012921/2010-...   

             first_name last_name                              birth_date  \
7769                NaN    Gerold           19.4. (Geburtsjahr unbekannt)   
11695            Johann    Kluser      wohl zweite Hälfte 14. Jahrhundert   
13809           Edouard   Matthey                                  ​15.6.   
14570               NaN   Moengal  Mitte bis zweite Hälfte 9. Jahrhundert   
16168  Giovan Francesco   Paleari                       2. Hälfte 15. Jh.   
18142               NaN   Romanus                         Ende des 4. Jh.   

                                        death_date  \
7769                                  10.4. um 978   
11695                                    nach 1437   
13809                                          NaN   
14570  an einem 30. September zwischen 868 und 895   
16168                                     vor 1542   
18142                                      460/465   

                                                    text   published  \
7769   19.4. (Geburtsjahr unbekannt), 10.4. um 978 . ...  29.08.2005   
11695  wohl zweite Hälfte 14. Jahrhundert, nach 1437....  14.12.2006   
13809  ​15.6.1833 Le Locle, ​23.7.1886 Prilly, reform...  08.04.2020   
14570  Mitte bis zweite Hälfte 9. Jahrhundert, an ein...  24.11.2008   
16168  2. Hälfte 15. Jh. Morcote, vor 1542 Morcote, k...  19.06.2009   
18142  Ende des 4. Jh., 460/465, ​La Balme (Saint-Rom...  10.11.2010   

                     author      translator   hhb_ids      hhb_forename  \
7769   Veronika Feller-Vest             NaN  11427839            Gerold   
11695          Hans Stadler             NaN  11447549            Johann   
13809       Marianne Derron             NaN       NaN               NaN   
14570           Peter Stotz             NaN  11429270           Moengal   
16168         Marino Viganò  Thomas Barfuss  11590730  Giovan Francesco   
18142  Veronika Feller-Vest             NaN  11429954           Romanus   

      hhb_surname hhb_sex hhb_birth_date hhb_death_date  hhb_relation_id  \
7769          NaN    male            NaN     0978-01-01              NaN   
11695      Kluser    male            NaN     1437-01-01              NaN   
13809         NaN     NaN            NaN            NaN              NaN   
14570         NaN    male            NaN            NaN              NaN   
16168     Paleari    male            NaN     1542-01-01              NaN   
18142         NaN    male            NaN     0460-01-01              NaN   

      hhb_relation_name hhb_relation_type  
7769                NaN               NaN  
11695               NaN               NaN  
13809               NaN               NaN  
14570               NaN               NaN  
16168               NaN               NaN  
18142               NaN               NaN

## Actual Cleaning of the birth date

In [18]:
df_scrape["cleaned_birth_date"] = df_scrape.birth_date.str.extract(r"((\d{1,2}\.\d{1,2}\.\d{3,4})|(\d{3,4}))")[0]
df_scrape.loc[11695, "cleaned_birth_date"] = "1.1.1375"
df_scrape.loc[13809, "cleaned_birth_date"] = "15.6.1833"
df_scrape.loc[14570, "cleaned_birth_date"] = "1.1.850"
df_scrape.loc[16168, "cleaned_birth_date"] = "1.1.1475"
df_scrape.loc[18142, "cleaned_birth_date"] = "1.1.400"

In [19]:
df_scrape["cleaned_birth_date"]

0          8.8.1599
1          7.4.1895
2         19.1.1894
3              1500
4          3.2.1898
            ...    
25266     24.1.1887
25267    11.11.1921
25268      7.3.1304
25269          1445
25270     15.6.1558
Name: cleaned_birth_date, Length: 25271, dtype: object

# modify cleaned birth date to be in the format YYYY-MM-DD (format="%Y-%m-%d")

In [20]:
# 8.8.1599 -> 1599-08-08
df_scrape["cleaned_birth_date_formatted"] = df_scrape[df_scrape["cleaned_birth_date"].str.fullmatch(r"(\d{1}\.\d{1}\.\d{4})", na=False)]["cleaned_birth_date"].str.replace(r"(\d{1})\.(\d{1})\.(\d{4})", r"\3-0\2-0\1", regex=True)
# 8.8.900 -> 0900-08-08
df_scrape["cleaned_birth_date_formatted"].fillna(df_scrape[df_scrape["cleaned_birth_date"].str.fullmatch(r"(\d{1}\.\d{1}\.\d{3})", na=False)]["cleaned_birth_date"].str.replace(r"(\d{1})\.(\d{1})\.(\d{3})", r"0\3-0\2-0\1", regex=True), inplace=True)
# 14.5.1900 -> 1900-05-14
df_scrape["cleaned_birth_date_formatted"].fillna(df_scrape[df_scrape["cleaned_birth_date"].str.fullmatch(r"(\d{2}\.\d{1}\.\d{4})", na=False)]["cleaned_birth_date"].str.replace(r"(\d{2})\.(\d{1})\.(\d{4})", r"\3-0\2-\1", regex=True), inplace=True)
# 14.5.900 -> 0900-05-14
df_scrape["cleaned_birth_date_formatted"].fillna(df_scrape[df_scrape["cleaned_birth_date"].str.fullmatch(r"(\d{2}\.\d{1}\.\d{3})", na=False)]["cleaned_birth_date"].str.replace(r"(\d{2})\.(\d{1})\.(\d{3})", r"0\3-0\2-\1", regex=True), inplace=True)
# 4.05.1900 -> 1900-05-04
df_scrape["cleaned_birth_date_formatted"].fillna(df_scrape[df_scrape["cleaned_birth_date"].str.fullmatch(r"(\d{1}\.\d{2}\.\d{4})", na=False)]["cleaned_birth_date"].str.replace(r"(\d{1})\.(\d{2})\.(\d{4})", r"\3-\2-0\1", regex=True), inplace=True)
# 4.05.500 -> 0500-05-04
df_scrape["cleaned_birth_date_formatted"].fillna(df_scrape[df_scrape["cleaned_birth_date"].str.fullmatch(r"(\d{1}\.\d{2}\.\d{3})", na=False)]["cleaned_birth_date"].str.replace(r"(\d{1})\.(\d{2})\.(\d{3})", r"0\3-\2-0\1", regex=True), inplace=True)
# 18.08.1500 -> 1500-08-18
df_scrape["cleaned_birth_date_formatted"].fillna(df_scrape[df_scrape["cleaned_birth_date"].str.fullmatch(r"(\d{2}\.\d{2}\.\d{4})", na=False)]["cleaned_birth_date"].str.replace(r"(\d{2})\.(\d{2})\.(\d{4})", r"\3-\2-\1", regex=True), inplace=True)
# 18.08.900 -> 0900-08-18
df_scrape["cleaned_birth_date_formatted"].fillna(df_scrape[df_scrape["cleaned_birth_date"].str.fullmatch(r"(\d{2}\.\d{2}\.\d{3})", na=False)]["cleaned_birth_date"].str.replace(r"(\d{2})\.(\d{2})\.(\d{3})", r"0\3-\2-\1", regex=True), inplace=True)
# 900 to 0900-01-01
df_scrape["cleaned_birth_date_formatted"].fillna(df_scrape[df_scrape["cleaned_birth_date"].str.fullmatch(r"(\d{3})", na=False)]["cleaned_birth_date"].str.replace(r"(\d{3})", r"0\1-01-01", regex=True), inplace=True)
# 1500 to 1500-01-01
df_scrape["cleaned_birth_date_formatted"].fillna(df_scrape[df_scrape["cleaned_birth_date"].str.fullmatch(r"(\d{4})", na=False)]["cleaned_birth_date"].str.replace(r"(\d{4})", r"\1-01-01", regex=True), inplace=True)

In [21]:
df_scrape["cleaned_birth_date_formatted"]

0        1599-08-08
1        1895-04-07
2        1894-01-19
3        1500-01-01
4        1898-02-03
            ...    
25266    1887-01-24
25267    1921-11-11
25268    1304-03-07
25269    1445-01-01
25270    1558-06-15
Name: cleaned_birth_date_formatted, Length: 25271, dtype: object

## Adding Missing Birth Dates from HHB

In [22]:
df_scrape.fillna({"cleaned_birth_date_formatted": df_scrape.hhb_birth_date}, inplace=True)

In [23]:
df_scrape["cleaned_birth_date_formatted"].isna().sum()

73

In [24]:
df_scrape[df_scrape["cleaned_birth_date_formatted"].isna()][["birth_date", "cleaned_birth_date", "hhb_birth_date", "cleaned_birth_date_formatted"]]

birth_date cleaned_birth_date hhb_birth_date  \
126                              NaN                NaN            NaN   
310                              NaN                NaN            NaN   
405                              NaN                NaN            NaN   
416                              NaN                NaN            NaN   
449                              NaN                NaN            NaN   
727                              NaN                NaN            NaN   
1023                             NaN                NaN            NaN   
1098                             NaN                NaN            NaN   
1497                             NaN                NaN            NaN   
1586                             NaN                NaN            NaN   
1631                             NaN                NaN            NaN   
1665                             NaN                NaN            NaN   
1670                             NaN                NaN            NaN   
1691                             NaN                NaN            NaN   
1848                             NaN                NaN            NaN   
2031                             NaN                NaN            NaN   
3177                             NaN                NaN            NaN   
3222                             NaN                NaN            NaN   
3303                             NaN                NaN            NaN   
5646                             NaN                NaN            NaN   
6147                             NaN                NaN            NaN   
6450                             NaN                NaN            NaN   
6888                             NaN                NaN            NaN   
6889                             NaN                NaN            NaN   
6893                             NaN                NaN            NaN   
7551                             NaN                NaN            NaN   
7769   19.4. (Geburtsjahr unbekannt)                NaN            NaN   
7774                             NaN                NaN            NaN   
7883                             NaN                NaN            NaN   
8197                             NaN                NaN            NaN   
8739                             NaN                NaN            NaN   
9049                             NaN                NaN            NaN   
10433                            NaN                NaN            NaN   
12064                            NaN                NaN            NaN   
12145                            NaN                NaN            NaN   
12477                            NaN                NaN            NaN   
12702                            NaN                NaN            NaN   
12808                            NaN                NaN            NaN   
14474                            NaN                NaN            NaN   
14988                            NaN                NaN            NaN   
15367                            NaN                NaN            NaN   
15440                            NaN                NaN            NaN   
16018                            NaN                NaN            NaN   
16189                            NaN                NaN            NaN   
16731                            NaN                NaN            NaN   
16787                            NaN                NaN            NaN   
16943                            NaN                NaN            NaN   
17171                            NaN                NaN            NaN   
17231                            NaN                NaN            NaN   
17294                            NaN                NaN            NaN   
17790                            NaN                NaN            NaN   
18029                            NaN                NaN            NaN   
18254                            NaN                NaN            NaN   
18531                         

## Death Date

In [25]:
df_scrape["cleaned_death_date"] = df_scrape.death_date.str.extract(r"((\d{1,2}\.\d{1,2}\.\d{3,4})|(\d{3,4}))")[0]
# 8.8.1599 -> 1599-08-08
df_scrape["cleaned_death_date_formatted"] = df_scrape[df_scrape["cleaned_death_date"].str.fullmatch(r"(\d{1}\.\d{1}\.\d{4})", na=False)]["cleaned_death_date"].str.replace(r"(\d{1})\.(\d{1})\.(\d{4})", r"\3-0\2-0\1", regex=True)
# 8.8.900 -> 0900-08-08
df_scrape["cleaned_death_date_formatted"].fillna(df_scrape[df_scrape["cleaned_death_date"].str.fullmatch(r"(\d{1}\.\d{1}\.\d{3})", na=False)]["cleaned_death_date"].str.replace(r"(\d{1})\.(\d{1})\.(\d{3})", r"0\3-0\2-0\1", regex=True), inplace=True)
# 14.5.1900 -> 1900-05-14
df_scrape["cleaned_death_date_formatted"].fillna(df_scrape[df_scrape["cleaned_death_date"].str.fullmatch(r"(\d{2}\.\d{1}\.\d{4})", na=False)]["cleaned_death_date"].str.replace(r"(\d{2})\.(\d{1})\.(\d{4})", r"\3-0\2-\1", regex=True), inplace=True)
# 14.5.900 -> 0900-05-14
df_scrape["cleaned_death_date_formatted"].fillna(df_scrape[df_scrape["cleaned_death_date"].str.fullmatch(r"(\d{2}\.\d{1}\.\d{3})", na=False)]["cleaned_death_date"].str.replace(r"(\d{2})\.(\d{1})\.(\d{3})", r"0\3-0\2-\1", regex=True), inplace=True)
# 4.05.1900 -> 1900-05-04
df_scrape["cleaned_death_date_formatted"].fillna(df_scrape[df_scrape["cleaned_death_date"].str.fullmatch(r"(\d{1}\.\d{2}\.\d{4})", na=False)]["cleaned_death_date"].str.replace(r"(\d{1})\.(\d{2})\.(\d{4})", r"\3-\2-0\1", regex=True), inplace=True)
# 4.05.500 -> 0500-05-04
df_scrape["cleaned_death_date_formatted"].fillna(df_scrape[df_scrape["cleaned_death_date"].str.fullmatch(r"(\d{1}\.\d{2}\.\d{3})", na=False)]["cleaned_death_date"].str.replace(r"(\d{1})\.(\d{2})\.(\d{3})", r"0\3-\2-0\1", regex=True), inplace=True)
# 18.08.1500 -> 1500-08-18
df_scrape["cleaned_death_date_formatted"].fillna(df_scrape[df_scrape["cleaned_death_date"].str.fullmatch(r"(\d{2}\.\d{2}\.\d{4})", na=False)]["cleaned_death_date"].str.replace(r"(\d{2})\.(\d{2})\.(\d{4})", r"\3-\2-\1", regex=True), inplace=True)
# 18.08.900 -> 0900-08-18
df_scrape["cleaned_death_date_formatted"].fillna(df_scrape[df_scrape["cleaned_death_date"].str.fullmatch(r"(\d{2}\.\d{2}\.\d{3})", na=False)]["cleaned_death_date"].str.replace(r"(\d{2})\.(\d{2})\.(\d{3})", r"0\3-\2-\1", regex=True), inplace=True)
# 900 to 0900-01-01
df_scrape["cleaned_death_date_formatted"].fillna(df_scrape[df_scrape["cleaned_death_date"].str.fullmatch(r"(\d{3})", na=False)]["cleaned_death_date"].str.replace(r"(\d{3})", r"0\1-01-01", regex=True), inplace=True)
# 1500 to 1500-01-01
df_scrape["cleaned_death_date_formatted"].fillna(df_scrape[df_scrape["cleaned_death_date"].str.fullmatch(r"(\d{4})", na=False)]["cleaned_death_date"].str.replace(r"(\d{4})", r"\1-01-01", regex=True), inplace=True)

# fill missing dates with hhd_death_date
df_scrape.fillna({"cleaned_death_date_formatted": df_scrape.hhb_death_date}, inplace=True)

In [26]:
df_scrape["cleaned_death_date_formatted"].isna().sum()

472

In [28]:
pd.options.display.max_rows = 500
df_scrape[df_scrape["cleaned_death_date_formatted"].isna()][["death_date", "cleaned_death_date", "hhb_death_date", "cleaned_death_date_formatted"]]

death_date cleaned_death_date hhb_death_date  \
122                                 NaN                NaN            NaN   
123                                 NaN                NaN            NaN   
130                                 NaN                NaN            NaN   
173                                 NaN                NaN            NaN   
336                                 NaN                NaN            NaN   
440                                 NaN                NaN            NaN   
511                                 NaN                NaN            NaN   
577                                 NaN                NaN            NaN   
633                                 NaN                NaN            NaN   
685                                 NaN                NaN            NaN   
700                                 NaN                NaN            NaN   
795                                 NaN                NaN            NaN   
797                                 NaN                NaN            NaN   
810                                 NaN                NaN            NaN   
811                                 NaN                NaN            NaN   
835                                 NaN                NaN            NaN   
838                                 NaN                NaN            NaN   
846                                 NaN                NaN            NaN   
1150                                NaN                NaN            NaN   
1218                                NaN                NaN            NaN   
1219                                NaN                NaN            NaN   
1296                                NaN                NaN            NaN   
1512                                NaN                NaN            NaN   
1532                                NaN                NaN            NaN   
1554                               19.1                NaN            NaN   
1682                                NaN                NaN            NaN   
1819                                NaN                NaN            NaN   
1842                                NaN                NaN            NaN   
1848                                NaN                NaN            NaN   
1854                                NaN                NaN            NaN   
1866                                NaN                NaN            NaN   
1886                                NaN                NaN            NaN   
1888                                NaN                NaN            NaN   
1926                                NaN                NaN            NaN   
1995                                NaN                NaN            NaN   
2031                                NaN                NaN            NaN   
2085                                NaN                NaN            NaN   
2158                                NaN                NaN            NaN   
2233                                NaN                NaN            NaN   
2415                                NaN                NaN            NaN   
2488                                NaN                NaN            NaN   
2490                                NaN                NaN            NaN   
2491                                NaN                NaN            NaN   
2530                                NaN                NaN            NaN   
2577                                NaN                NaN            NaN   
2598                                NaN                NaN            NaN   
2600                                NaN                NaN            NaN   
2663                                NaN                NaN            NaN   
2723                                NaN                NaN            NaN   
2939                                NaN                NaN            NaN   
3021                                NaN                NaN            NaN   
3034                     

In [31]:
df_scrape.drop("cleaned_birth_date",axis=1, inplace=True)
df_scrape.drop("cleaned_death_date",axis=1, inplace=True)

In [32]:
df_scrape.to_csv("hls_dump/22_10_22_hls_people_hist_hub_cleaned.csv", sep=";", index=False)

HLS_ID                                            HLS_URL  \
0       25265  http://hls-dhs-dss.ch/de/articles/025265/2001-...   
1       27932  http://hls-dhs-dss.ch/de/articles/027932/2002-...   
2        4928  http://hls-dhs-dss.ch/de/articles/004928/2002-...   
3       11462  http://hls-dhs-dss.ch/de/articles/011462/2001-...   
4        8115  http://hls-dhs-dss.ch/de/articles/008115/2001-...   
...       ...                                                ...   
25266   27791  http://hls-dhs-dss.ch/de/articles/027791/2013-...   
25267   30267  http://hls-dhs-dss.ch/de/articles/030267/2013-...   
25268   21194  http://hls-dhs-dss.ch/de/articles/021194/2008-...   
25269   18772  http://hls-dhs-dss.ch/de/articles/018772/2009-...   
25270   26337  http://hls-dhs-dss.ch/de/articles/026337/2001-...   

           first_name   last_name  birth_date  death_date  \
0       Johann Ludwig     A Musis    8.8.1599   10.6.1662   
1      Carlo Emanuele     A Prato    7.4.1895   17.9.1968   
2      Albert von der          Aa   19.1.1894   10.7.1978   
3            Johannes         Aal     um 1500   28.5.1551   
4               Alvar       Aalto    3.2.1898   11.5.1976   
...               ...         ...         ...         ...   
25266        Giovanni        Züst   24.1.1887   4.12.1976   
25267       Hansruedi        Züst  11.11.1921  22.11.2005   
25268      Rudolf von    Ödisried    7.3.1304        1332   
25269          Gallus        Öhem     um 1445        1522   
25270     Andreas von  Österreich   15.6.1558  12.11.1600   

                                                    text   published  \
0      8.8.1599 Pfullendorf (Baden, D), 10.6.1662 Sol...  10.05.2001   
1      7.4.1895 Trient, 17.9.1968 Genf, ursprünglich ...  21.08.2002   
2      19.1.1894 Vevey, 10.7.1978 Lausanne, reformier...  11.06.2002   
3      um 1500 Bremgarten (AG), 28.5.1551 Solothurn, ...  24.01.2001   
4      3.2.1898 Kuortane (Finnland), 11.5.1976 Helsin...  15.05.2001   
...                                                  ...         ...   
25266  24.1.1887 Basel, 4.12.1976 Rancate (heute Geme...  24.04.2013   
25267  11.11.1921 Rheineck, 22.11.2005 Zürich, reform...  28.01.2013   
25268  Erstmals erwähnt 7.3.1304, 1332 oder kurz dana...  04.11.2008   
25269  um 1445 Radolfzell, 1522 Freiburg i.Br. Illegi...  20.08.2009   
25270  15.6.1558 Burg Bresnitz (Böhmen), 12.11.1600 R...  21.11.2001   

                          author                translator  ...  \
0                     Urban Fink                       NaN  ...   
1                  Mauro Cerutti  Christoph Neuenschwander  ...   
2                 Gilbert Marion               Peter Opitz  ...   
3                 Rolf Max Kully                       NaN  ...   
4               Simona Martinoli              Sabine Kraut  ...   
...                          ...                       ...  ...   
25266  Mariangela Agliati Ruggia            Christa Mathys  ...   
25267            Adrian Scherrer                       NaN  ...   
25268           Niklaus von Flüe                       NaN  ...   
25269       Veronika Feller-Vest                       NaN  ...   
25270               Herbert Frey                       NaN  ...   

         hhb_forename     hhb_surname hhb_sex hhb_birth_date hhb_death_date  \
0       Johann Ludwig         A Musis    male     1599-08-08     1662-06-10   
1      Carlo Emanuele         A Prato    male     1895-04-07     1968-09-17   
2              Albert      von der Aa    male     1894-01-19     1978-07-10   
3            Johannes             Aal    male     1500-01-01     1551-05-28   
4               Alvar           Aalto    male     1898-02-03     1976-05-11   
...               ...             ...     ...            ...            ...   
25266        Giovanni            Züst    male     1887-01-24     1976-12-04   
25267       Hansruedi            Züst    male     1921-11-11     2005-11-22   
25268          Rudolf    von Ödisried    male     1304-01-01     1332-01-01 